### DISCUSSION TITLE
Read markdown page frontmatter without rendering the page

### DISCUSSION THREAD

**Discussion by tony-sull** [withastro/roadmap#118] [Proposal]
I'd like to be able to read the frontmatter from local markdown pages without actually rendering those pages' HTML content

## Use case

I want to have a side panel component with navigation links to every page in my project, similar to the nav menu used in Astro's [docs](https://docs.astro.build) site.

### The challenge

`Astro.fetchContent()` is currently the best way to load metadata for local `*.md` files, but it's easy to accidentally create a circular loop that hangs the build.  For example, if the side panel component is used in a shared layout, the `*.md` files would also load that layout and call `fetchContent` again

From a performance perspective, it'd also be nice to know the build won't waste any time rendering HTML content for each page when I really just want frontmatter data

## Possible solutions (WIP)

- Create a new `Astro.fetchFrontmatter()` API that works much like `fetchContent()` does today, but always ignores the markdown file's `layout` frontmatter and doesn't try to render the page's content

- Update `Astro.fetchContent()` with the functionality above.  ⚠️ This would be a breaking change, and it does kind of make sense that an API named `fetchContent` would try to render the page's content

- Expose something similar to the sitemap on the `Astro` global

**Comment by FredKSchott** [comment]
https://discord.com/channels/830184174198718474/845430950191038464/937257754073128980

Copying a comment I made in Discord, experimenting with this: 

an experiment: I was able to speed up the docs dev startup by ~70% with a Rollup trick: 

1. Markdown plugin loads the `.md` file, does the raw frontmatter + MD->HTML parse. Returns a wrapper that, on render, does a dynamic import to a virtual module.
2. The `.md` -> `.astro` transform is left for the virtual module. It only runs when the render is actually used. `fetchContent()` is super fast now, since you only pay the full render cost on actual usage.

```js
`export const frontmatter = ${JSON.stringify(frontmatter)};

export default async function render(...args) {
  return (await import(${JSON.stringify(cacheId)})).default(...args);
};`
```

experiment lives here: https://github.com/withastro/astro/commit/d3d1f994f23b562206435715169bc1ad3c8c16aa

I'm not sure if this helps our build times or the benchmark that you're working towards, but I found the trick interesting and was surprised how well it worked to speed up dev.

**Reply by FredKSchott** [reply]
This discord comment was about perf, but I just realized it would solve our infinite loop problem as well and address this discussion as a result. Maybe worth pursuing? Would be fun to pair on this!

**Reply by songololo** [reply]
I've been running into really slow page refreshes during development and now realise it is probably to do with this issue...!

**Reply by ghost** [reply]
This is how 11ty handles collections! Each item in the collection returns a callback for `renderTemplate`, so costs are only incurred on usage. It'd be super convenient if Astro did the same 👍

**Comment by aFuzzyBear** [comment]
Hey Tony, this is how I was shown to over come this, dont know if it is right:
```jsx

const paths = Object.keys(import.meta.glob('../path/to/markdown/content/**/*.md'))
    
function globby(){
    const docdir = '../path/to/markdown/content/'
    return paths.map(
        (slug) => {
         return ({
           params: {
               slug: slug.slice(docdir.length,-3).split('/').slice(1).join('/').toLowerCase(),
           },
           props: {
               slug: slug.slice(docdir.length,-3).split('/').slice(1).join('/').toLowerCase(),
               path: new URL(slug, import.meta.url).pathname,
           },
       })}
   ).filter(
          x=>!(x.props.directory.includes('readme'))
           )
}

const fetchData = async(path)=>{
    return await import(path).then(exports=>exports.frontmatter)
  }
  /** Attaches frontmatter and data from each file alongside nested menu of static paths. */
  const withMetaData = async (paths) => {
      const hash = Object.create(null)
      for (const path of paths) {
          const data = await fetchData(path.props.path)
          const metaData = Object.entries(data)
              .filter(result=>!result.includes('astro'))
              .reduce((prev,current)=>({...prev,[current[0]]:current[1]}),{})
          path.props.data = data
          paths.props.metaData = metaData
          }
   }
   export default async function(){ return await withMetaData(globby())}
   ```

**Comment by songololo** [comment]
I've also run into the circular dependency issue when trying to implement a sidebar for navigation.

What is the purpose of rendering the layout when running `fetchContent`? (Or is this used internally to render the pages?)

In my case (and I imagine most end-user cases?) I simply want to load the front matter and parsed markdown content, not render the final markdown page as it would render to the final linked layout.

Is there a suggested workaround for this in the meantime?

**Comment by songololo** [comment]
Adding similar discussions and PRs:
https://github.com/withastro/astro/issues/1247
https://github.com/withastro/rfcs/discussions/5
https://github.com/withastro/astro/issues/1895
https://github.com/withastro/astro/pull/2657

**Comment by DaveOrDead** [comment]
I'm in the same boat as well. The site I'm working on has both a Blog area and a Documentation area that have both run into the fetchContent infinite loop / circular dependency  issue but only when building - dev mode is fine.

In my setup both areas have a collection page listing out the articles and then:
a) Doc pages have a navigation sidebar containing articles from the same directory
b) Blog area has related articles list at the bottom of a post

My current workaround has just been to have separate JSON file that mirrors the site structure / front matter but this obviously doesn't scale and will quickly become out of sync.

Keen to know if others have a smarter solution than mine as a workaround?

**Reply by songololo** [reply]
Another option:
```
const allMd = import.meta.glob('./pages/**/*.md', { assert: { type: 'raw' } })
```
